### ISSUE TITLE
job-manager: Incorrect exception cb request issuance?

### ISSUE THREAD

**Issue by dongahn** [flux-framework/flux-core#2220]
I used the "single" queue depth policy so that flux-sched gets one alloc request at a time. 

I submitted more jobs than the scheduler can schedule and then canceled those jobs whose alloc requests have not even seen by the scheduler. But it appears that `flux job cancel` on those "unseen" jobs still lead to an invocation of the exception callback within `qmanager`. 

I can gracefully handle these "nonexistent" jobs from within the exception callback with `ENOENT` but it would be probably better if `jobmanager` doesn't raise an exception to the scheduler in this case.

I can schedule 4 single core jobs on my Mac OSx docker. I submitted 6 jobs:

```console
ahn1@da15834410b5:/usr/src$ flux start -s 1 -o,-S,log-filename=out
flux-start: warning: setting --bootstrap=selfpmi due to --size option
To run a command as administrator (user "root"), use "sudo <command>".
See "man sudo_root" for details.

ahn1@da15834410b5:/usr/src$
ahn1@da15834410b5:/usr/src$
ahn1@da15834410b5:/usr/src$ flux job submit test.json
152706220032
ahn1@da15834410b5:/usr/src$ flux job submit test.json
162588000256
ahn1@da15834410b5:/usr/src$ flux job submit test.json
170775281664
ahn1@da15834410b5:/usr/src$ flux job submit test.json
178157256704
ahn1@da15834410b5:/usr/src$ flux job submit test.json
186562641920
ahn1@da15834410b5:/usr/src$ flux job submit test.json
197065179136
ahn1@da15834410b5:/usr/src$
ahn1@da15834410b5:/usr/src$ flux job list
JOBID		STATE	USERID	PRI	T_SUBMIT
152706220032	R	7818	16	2019-07-10T04:24:02Z
162588000256	R	7818	16	2019-07-10T04:24:02Z
170775281664	R	7818	16	2019-07-10T04:24:03Z
178157256704	R	7818	16	2019-07-10T04:24:03Z
186562641920	S	7818	16	2019-07-10T04:24:04Z
197065179136	S	7818	16	2019-07-10T04:24:04Z
ahn1@da15834410b5:/usr/src$ flux job cancel 197065179136
2019-07-10T04:24:32.969590Z qmanager.err[0]: jobmanager_exception_cb: remove: No such file or directory
ahn1@da15834410b5:/usr/src$
ahn1@da15834410b5:/usr/src$ flux job list
JOBID		STATE	USERID	PRI	T_SUBMIT
152706220032	R	7818	16	2019-07-10T04:24:02Z
162588000256	R	7818	16	2019-07-10T04:24:02Z
170775281664	R	7818	16	2019-07-10T04:24:03Z
178157256704	R	7818	16	2019-07-10T04:24:03Z
186562641920	S	7818	16	2019-07-10T04:24:04Z
ahn1@da15834410b5:/usr/src$ flux job cancel 186562641920
ahn1@da15834410b5:/usr/src$
ahn1@da15834410b5:/usr/src$ flux job list
JOBID		STATE	USERID	PRI	T_SUBMIT
152706220032	R	7818	16	2019-07-10T04:24:02Z
162588000256	R	7818	16	2019-07-10T04:24:02Z
170775281664	R	7818	16	2019-07-10T04:24:03Z
178157256704	R	7818	16	2019-07-10T04:24:03Z
```

**Comment by garlick** [comment]
Right now, for simplicity, we are publishing job exceptions as open-loop event messages and libschedutil will pass all of these on to the scheduler.

I think sched should ignore all exceptions except those for jobs that have an outstanding alloc request that has not been fulfilled (responded to) and severity of zero (fatal).

If this design is ultimately not a good fit we could do it another way, but I do sort of like the idea of exceptions as a published event that can be consumed by any part of the system.

See also
https://github.com/flux-framework/rfc/blob/master/spec_21.adoc#exceptions

**Comment by grondo** [comment]
Perhaps open an issue against schedutil to optionally filter these exceptions since all schedulers will need to do this anyway (avoids code duplication, though not much)?

**Comment by dongahn** [comment]
@garlick: Ok. Somehow I didn't realize this was implemented using the event.

> Perhaps open an issue against schedutil to optionally filter these exceptions since all schedulers will need to do this anyway (avoids code duplication, though not much)?

Like I said above, I already handle this pretty efficiently at the scheduler level -- (`O(log n)`) -- given how I implemented the queues. So, this shouldn't be necessary for my case. In fact, I am a bit worried about this approach, as to filter these, schedutil will have to have some job indexing etc which can increase both performance and memory overhead.

I am closing this issue. It was good to log this here if this behavior surprises others, though.
